x-pack/filebeat/input/awss3/sqs.go

```diff
@@ -76,37 +158,88 @@ func (r *sqsReader) Receive(ctx context.Context) error {
 		// Release unused workers.
 		r.workerSem.Release(workers - len(msgs))
 
-		// Process each SQS message asynchronously with a goroutine.
 		r.log.Debugf("Received %v SQS messages.", len(msgs))
 		r.metrics.sqsMessagesReceivedTotal.Add(uint64(len(msgs)))
-		workerWg.Add(len(msgs))
+
+		// Add to processing wait group to wait for all messages to be processed.
+		processingWg.Add(len(msgs))
 
 		for _, msg := range msgs {
+			// Process each SQS message asynchronously with a goroutine.
 			go func(msg types.Message, start time.Time) {
 				id := r.metrics.beginSQSWorker()
 				defer func() {
+					// Mark processing wait group as done.
 					r.metrics.endSQSWorker(id)
-					workerWg.Done()
+					processingWg.Done()
 					r.workerSem.Release(1)
 				}()
 
-				if err := r.msgHandler.ProcessSQS(ctx, &msg); err != nil {
+				// Create a pipeline client scoped to this goroutine.
+				client, err := pipeline.ConnectWith(beat.ClientConfig{
+					EventListener: awscommon.NewEventACKHandler(),
+					Processing: beat.ProcessingConfig{
+						// This input only produces events with basic types so normalization
+						// is not required.
+						EventNormalization: boolPtr(false),
+					},
+				})
+
+				if err != nil {
 					r.log.Warnw("Failed processing SQS message.",
 						"error", err,
 						"message_id", *msg.MessageId,
 						"elapsed_time_ns", time.Since(start))
+
+					return
+				}
+
+				defer client.Close()
+
+				r.log.Debugw("Going to process SQS message.",
+					"message_id", *msg.MessageId,
+					"elapsed_time_ns", time.Since(start))
+
+				acker := awscommon.NewEventACKTracker()
```

Comment:
`acker` initialization is moved here and extracted as dependency injection mostly for testing purpose.
I cannot think of a clean workaround to avoid extraction of the initialization for testing purpose